packages/block-editor/src/components/block-list/use-in-between-inserter.js

```diff
@@ -68,26 +69,43 @@ export function useInBetweenInserter() {
 					rootClientId = blockElement.getAttribute( 'data-block' );
 				}
 
+				const position = { x: event.clientX, y: event.clientY };
 				const orientation =
 					getBlockListSettings( rootClientId )?.orientation ||
 					'vertical';
-				const rect = event.target.getBoundingClientRect();
-				const offsetTop = event.clientY - rect.top;
-				const offsetLeft = event.clientX - rect.left;
 
 				const children = Array.from( event.target.children );
-				let element = children.find( ( blockEl ) => {
-					return (
-						( blockEl.classList.contains( 'wp-block' ) &&
-							orientation === 'vertical' &&
-							blockEl.offsetTop > offsetTop ) ||
-						( blockEl.classList.contains( 'wp-block' ) &&
-							orientation === 'horizontal' &&
-							blockEl.offsetLeft > offsetLeft )
+				let element;
+				let elementDistance;
+				let indexOffset;
+
+				children.forEach( ( blockEl ) => {
+					const allowedEdges =
+						orientation === 'horizontal'
+							? [ 'left', 'right' ]
+							: [ 'top', 'bottom' ];
+					const elementRect = blockEl.getBoundingClientRect();
+					const [ distance, edge ] = getDistanceToNearestEdge(
+						position,
+						elementRect,
+						allowedEdges
 					);
+
+					if ( ! elementDistance || distance < elementDistance ) {
+						element = blockEl;
+						elementDistance = distance;
+						// Use the next index if the cursor is closest to the
+						// trailing end of the block.
+						indexOffset = [ 'right', 'bottom' ].includes( edge )
+							? 1
+							: 0;
+					}
 				} );
 
 				if ( ! element ) {
+					if ( isBlockInsertionPointVisible() ) {
```

Comment:
Not sure if this change is right, I wasn't fully able to understand the original code.

Reply:
Seems good to me :)

Reply:
We should probably be adding it to all early returns below.